File: frontend/src/lib/cardValidation.ts
import type { CardBrand } from '../types'

const BRAND_RULES: { brand: CardBrand; prefixes: string[]; lengths: number[] }[] = [
  { brand: 'visa', prefixes: ['4'], lengths: [13, 16, 19] },
  {
    brand: 'mastercard',
    prefixes: [
      ...Array.from({ length: 5 }, (_, i) => String(51 + i)),
      ...Array.from({ length: 500 }, (_, i) => String(2221 + i)),
    ],
    lengths: [16],
  },
  {
    brand: 'discover',
    prefixes: ['6011', '65', ...Array.from({ length: 6 }, (_, i) => String(644 + i))],
    lengths: [16],
  },
  { brand: 'amex', prefixes: ['34', '37'], lengths: [15] },
]

function digitsOnly(value: string): string {
  return value.replace(/[^0-9]/g, '')
}

export function detectBrand(number: string): CardBrand | null {
  const digits = digitsOnly(number)
  for (const rule of BRAND_RULES) {
    if (!rule.lengths.includes(digits.length)) continue
    if (rule.prefixes.some((prefix) => digits.startsWith(prefix))) return rule.brand
  }
  return null
}

export function luhnIsValid(number: string): boolean {
  const digits = digitsOnly(number).split('').map(Number)
  if (digits.length === 0) return false
  let checksum = 0
  digits.reverse().forEach((digit, index) => {
    let value = digit
    if (index % 2 === 1) {
      value *= 2
      if (value > 9) value -= 9
    }
    checksum += value
  })
  return checksum % 10 === 0
}

export function cvvLengthFor(brand: CardBrand): number {
  return brand === 'amex' ? 4 : 3
}

interface CardInput {
  brand: CardBrand
  number: string
  expMonth: number
  expYear: number
  cvv: string
  postalCode: string
}

/** Field name -> error message. Empty object means the card looks valid. */
export function validateCard(input: CardInput): Record<string, string> {
  const errors: Record<string, string> = {}
  const digits = digitsOnly(input.number)

  if (digits.length < 12 || digits.length > 19) {
    errors.number = 'Card number length looks wrong.'
  } else if (!luhnIsValid(digits)) {
    errors.number = 'Card number failed the checksum check.'
  } else if (detectBrand(digits) !== input.brand) {
    errors.number = `That number doesn't look like a ${input.brand} card.`
  }

  if (input.expMonth < 1 || input.expMonth > 12) {
    errors.exp_month = 'Expiry month must be between 1 and 12.'
  } else {
    const now = new Date()
    const lastDayOfExpiryMonth = new Date(input.expYear, input.expMonth, 0)
    if (lastDayOfExpiryMonth < new Date(now.getFullYear(), now.getMonth(), now.getDate())) {
      errors.exp_year = 'This card has already expired.'
    }
  }

  const cvvDigits = digitsOnly(input.cvv)
  if (cvvDigits.length !== cvvLengthFor(input.brand)) {
    errors.cvv = `${input.brand} security codes are ${cvvLengthFor(input.brand)} digits.`
  }

  if (!/^\d{5}(-\d{4})?$/.test(input.postalCode)) {
    errors.postal_code = 'Zip code must be 5 digits (optionally +4).'
  }

  return errors
}

File: backend/app/services/card_validation.py
"""Card-field validation. Pure functions: no I/O, no DB, no Stripe.

Only ever handles a card number/CVV in memory to validate it — nothing here
persists them. Callers must discard the raw values after this returns.
"""

from __future__ import annotations

import calendar
import re
from dataclasses import dataclass
from datetime import date

_DIGITS_ONLY = re.compile(r"[^0-9]")

_BRAND_RULES: dict[str, tuple[tuple[str, ...], tuple[int, ...]]] = {
    "visa": (("4",), (13, 16, 19)),
    "mastercard": (
        tuple(str(prefix) for prefix in range(51, 56))
        + tuple(str(prefix) for prefix in range(2221, 2721)),
        (16,),
    ),
    "discover": (("6011", "65") + tuple(str(prefix) for prefix in range(644, 650)), (16,)),
    "amex": (("34", "37"), (15,)),
}


@dataclass(frozen=True)
class FieldError:
    """One validation failure, tied to the input field that caused it."""

    field: str
    message: str


def _clean_number(number: str) -> str:
    return _DIGITS_ONLY.sub("", number)


def detect_brand(number: str) -> str | None:
    """Guess the card network from the number's prefix and length."""
    digits = _clean_number(number)
    for brand, (prefixes, lengths) in _BRAND_RULES.items():
        if len(digits) not in lengths:
            continue
        if any(digits.startswith(prefix) for prefix in prefixes):
            return brand
    return None


def luhn_is_valid(number: str) -> bool:
    """Standard mod-10 checksum."""
    digits = [int(char) for char in _clean_number(number)]
    if not digits:
        return False
    checksum = 0
    for index, digit in enumerate(reversed(digits)):
        if index % 2 == 1:
            digit *= 2
            if digit > 9:
                digit -= 9
        checksum += digit
    return checksum % 10 == 0


def cvv_length_for(brand: str) -> int:
    """American Express uses a 4-digit CVV; every other network uses 3."""
    return 4 if brand == "amex" else 3


def validate_card(
    *,
    brand: str,
    number: str,
    exp_month: int,
    exp_year: int,
    cvv: str,
    postal_code: str,
    today: date,
) -> list[FieldError]:
    """Run every check and return every failure (not just the first)."""
    errors: list[FieldError] = []
    digits = _clean_number(number)

    if len(digits) < 12 or len(digits) > 19:
        errors.append(FieldError("number", "Card number length looks wrong."))
    elif not luhn_is_valid(digits):
        errors.append(FieldError("number", "Card number failed the checksum check."))
    elif detect_brand(digits) != brand:
        errors.append(
            FieldError("number", f"That number doesn't look like a {brand} card.")
        )

    if exp_month < 1 or exp_month > 12:
        errors.append(FieldError("exp_month", "Expiry month must be between 1 and 12."))
    else:
        last_day = calendar.monthrange(exp_year, exp_month)[1]
        expiry = date(exp_year, exp_month, last_day)
        if expiry < today:
            errors.append(FieldError("exp_year", "This card has already expired."))

    cvv_digits = _DIGITS_ONLY.sub("", cvv)
    if len(cvv_digits) != cvv_length_for(brand):
        errors.append(
            FieldError("cvv", f"{brand.title()} security codes are {cvv_length_for(brand)} digits.")
        )

    if not re.fullmatch(r"\d{5}(-\d{4})?", postal_code):
        errors.append(FieldError("postal_code", "Zip code must be 5 digits (optionally +4)."))

    return errors
